frontend/widgets/AudioMixer.cpp

```diff
@@ -0,0 +1,1035 @@
+		layoutButton->setIcon(layoutIcon);
+	}
+
+	QWidget *buttonWidget = mixerToolbar->widgetForAction(layoutButton);
+	if (buttonWidget) {
+		idian::Utils::toggleClass(buttonWidget, "icon-layout-vertical", vertical);
+		idian::Utils::toggleClass(buttonWidget, "icon-layout-horizontal", !vertical);
+	}
+}
+
+void AudioMixer::showMixerContextMenu()
+{
+	QAction unhideAllAction(QTStr("UnhideAll"), &mixerMenu);
+
+	QWidgetAction showHiddenAction(&mixerMenu);
+	MenuCheckBox showHiddenCheckBox(QTStr("Basic.AudioMixer.ShowHidden"), &mixerMenu);
+	showHiddenCheckBox.setAction(&showHiddenAction);
+	showHiddenCheckBox.setChecked(showHidden);
+	showHiddenAction.setDefaultWidget(&showHiddenCheckBox);
+
+	QWidgetAction showInactiveAction(&mixerMenu);
+	MenuCheckBox showInactiveCheckBox(QTStr("Basic.AudioMixer.ShowInactive"), &mixerMenu);
+	showInactiveCheckBox.setAction(&showInactiveAction);
+	showInactiveCheckBox.setChecked(showInactive);
+	showInactiveAction.setDefaultWidget(&showInactiveCheckBox);
+
+	QWidgetAction hiddenOnRightAction(&mixerMenu);
+	const char *hiddenShifted = mixerVertical ? "Basic.AudioMixer.KeepHiddenRight"
+						  : "Basic.AudioMixer.KeepHiddenBottom";
+	MenuCheckBox hiddenOnRightCheckBox(QTStr(hiddenShifted), &mixerMenu);
+	hiddenOnRightCheckBox.setAction(&hiddenOnRightAction);
+	hiddenOnRightCheckBox.setChecked(keepHiddenRight);
+	hiddenOnRightAction.setDefaultWidget(&hiddenOnRightCheckBox);
+
+	QWidgetAction inactiveOnRightAction(&mixerMenu);
+	const char *inactiveShifted = mixerVertical ? "Basic.AudioMixer.KeepInactiveRight"
+						    : "Basic.AudioMixer.KeepInactiveBottom";
+	MenuCheckBox inactiveOnRightCheckBox(QTStr(inactiveShifted), &mixerMenu);
+	inactiveOnRightCheckBox.setAction(&inactiveOnRightAction);
+	inactiveOnRightCheckBox.setChecked(keepInactiveRight);
+	inactiveOnRightAction.setDefaultWidget(&inactiveOnRightCheckBox);
+
+	/* ------------------- */
+
+	connect(&unhideAllAction, &QAction::triggered, this, &AudioMixer::unhideAllAudioControls, Qt::DirectConnection);
+
+	connect(&showHiddenCheckBox, &QCheckBox::toggled, this, &AudioMixer::toggleShowHidden, Qt::DirectConnection);
+	connect(&hiddenOnRightCheckBox, &QCheckBox::toggled, this, &AudioMixer::toggleKeepHiddenRight,
+		Qt::DirectConnection);
+
+	connect(&showInactiveCheckBox, &QCheckBox::toggled, this, &AudioMixer::toggleShowInactive,
+		Qt::DirectConnection);
+	connect(&inactiveOnRightCheckBox, &QCheckBox::toggled, this, &AudioMixer::toggleKeepInactiveRight,
+		Qt::DirectConnection);
+
+	/* ------------------- */
+
+	mixerMenu.addAction(&unhideAllAction);
+	mixerMenu.addSeparator();
+	mixerMenu.addAction(&showHiddenAction);
+	mixerMenu.addAction(&showInactiveAction);
+	mixerMenu.addAction(&hiddenOnRightAction);
+	mixerMenu.addAction(&inactiveOnRightAction);
+	mixerMenu.exec(QCursor::pos());
+}
+
+void AudioMixer::addControlForUuid(QString uuid)
+{
+	OBSSourceAutoRelease source = obs_get_source_by_uuid(uuid.toStdString().c_str());
```

Comment:
You're working with a C API here, which means you need to juggle raw C types again and cannot entire rely on "good" C++ types.

In this case I'd argue that using Qt's own conversion to a UTF-8 char array is sufficient:

```suggestion
	OBSSourceAutoRelease source = obs_get_source_by_uuid(uuid.toUtf8().constData());
```
I could imagine that Qt will fill a `QByteArray` with the UTF-8 bytes from the original UTF-16 string (first copy) and that `constData()` shares a pointer to that byte array, whereas `toStdString` will create a `std::string` instance with those same bytes copied over (a second copy) before sharing the underlying pointer to _those_ bytes.

I'd argue that a single conversion is sufficient here, particularly as you need to provide low-level C data to a C API.